// decentraland-transactions/errors.js
export var ErrorCode;
(function (ErrorCode) {
    ErrorCode["UNKNOWN"] = "unknown";
    ErrorCode["DAPP_LIMIT_REACHED"] = "dapp_limit_reached";
    ErrorCode["USER_LIMIT_REACHED"] = "user_limit_reached";
    ErrorCode["API_LIMIT_REACHED"] = "api_limit_reached";
    ErrorCode["GAS_LIMIT_REACHED"] = "gas_limit_reached";
    ErrorCode["EXPECTATION_FAILED"] = "expectation_failed";
    ErrorCode["INVALID_TRANSACTION"] = "invalid_transaction";
    ErrorCode["INVALID_SCHEMA"] = "invalid_schema";
    ErrorCode["INVALID_CONTRACT_ADDRESS"] = "invalid_contract_address";
    ErrorCode["INVALID_ADDRESS"] = "invalid_address";
    ErrorCode["SALE_PRICE_TOO_LOW"] = "sale_price_too_low";
    ErrorCode["QUOTA_REACHED"] = "quota_reached";
    ErrorCode["CONTRACT_ACCOUNT"] = "contract_account";
    ErrorCode["HIGH_CONGESTION"] = "high_congestion";
    ErrorCode["USER_DENIED"] = "user_denied";
})(ErrorCode || (ErrorCode = {}));
export class MetaTransactionError extends Error {
    constructor(message, code) {
        super(message);
        this.code = code;
        // Set the prototype explicitly.
        Object.setPrototypeOf(this, MetaTransactionError.prototype);
    }
}
